### ISSUE TITLE
Understanding flow to load the database with past years info

### ISSUE THREAD

**Issue by jbholden** [blreams/collegefootballpick10#10]
I was wondering if you could describe your ideas on how you plan to load the data from the spreadsheet into the database.  I see some code in stage_history.py, stage_models.py, and excel_history/excel/spreadsheet_test.py to load info into the database.  Is your intention to just run one of these scripts to load all of the data into the database?

In the cdcpool google database, I implemented this behavior by doing something like following.
```
1.  Expose the database models thru a custom HTTP REST API framework
2.  Create a script (i.e. scripts/excel directory) to read info from excel files
3.  Create a script (i.e scripts/api directory) to call the HTTP REST API
4.  Create a script (i.e. scripts/fbpool directory) that implements commands to load the database
5.  Run fbpool script commands to load the database.  This script will read from the excel file 
and use the API calls to add data to the database. 
```

The file https://github.com/jbholden/cdcpool_google/blob/master/scripts/fbpool/README.md provides
more detail and some examples.

**Comment by blreams** [comment]
Loading history into the DB from spreadsheets is a one-time deal. IOW, you do it once and then don't ever have to do it again. At least that is true once the DB models are in their final form...I think we are still doing some back and forth on exactly what the DB models will be in their final form. And as long as the DB models are changing, the code that is used to load history into the DB will also be changing to match.

Do I think we need something more (like the custom REST API framework)? I'm not sure that it is worth the additional effort, because you won't be loading history all that often (ideally, only once). There is a related issue of how the DB is backed up. Once the site is up and running there should never be a need to go all the way back to the history load from spreadsheets. I guess my view is that backing up the DB is external to the website side of things. How often the DB is backed up and where the backed up data is stored is part of that discussion.

My initial thoughts were to create stage_history.py to act as a standalone script that would load the history from spreadsheets into the DB. It is probably worth implementing a command line argument allowing you to specify a list of years to load.

stage_models.py
The intent for this was when I first started coding pick10/models.py, I wanted a way to pre-load the DB models with some data that I could use for testing purposes, nothing more. In fact any data that is loaded using stage_models.py should be considered extraneous data.

stage_history.py
The intent for this was to parse data from the history spreadsheets and build the actual DB models using that data. But it was always intended to be run standalone.

excel_history
This started out as a direct copy of the same from the google appengine framework ("stolen with pride"). Since you had already implemented much of what was needed to pull the requisite data from the spreadsheets, I figured why not. I only made a few changes for compatibility with django (as near as I can remember).

So, as stage_history.py becomes more and more complete for pulling in whole years of data from the spreadsheets, more of the tools from excel_history will be needed.

One thing that I have noticed, stage_history.py is not entirely standalone. I think you actually have to run stage_models.py first, then you can run stage_history.py. That should NOT be the case, and I am going to figure out what needs to change in stage_history to removed that dependence. I'm going to open new Issue to track that and close it once I have committed the fix.

**Comment by blreams** [comment]
I made some improvements to stage_history.py. Beware, it can take a while to run if you are starting your database from scratch. It does try to detect the existence of db models, and skips loading them if they are already present.

I added enough code that it is possible to load the entire 18 year history of the pool. This can take a long time. Ironically, this actually run fastest on my laptop in cygwin. Here are some subjective times for loading various parts of the database:

  * Loading all users takes just a few seconds.
  * Loading all teams/conferences takes seconds to tens of seconds.
  * Loading one year of games takes about 20 seconds.
     * Loading one week of one year of games takes one or two seconds.
  * Loading one year of picks takes roughly 2 minutes.
     * Loading one week of one year of picks takes about 8 seconds.

All that adds up to something on the order of 30 minutes to create the entire pool history from scratch.

I've also been fixing spreadsheet discrepancies as I find them. For example, my last commit included pool_2014_standings.xlsm because I noticed that year 2014, Wk09, Player Dave Locklear, Game 10 had data in both cells. There have been a number of other fixes to other year spreadsheets as well.

**Comment by blreams** [comment]
I'm leaving this open as an issue (w/ a question label). If you think that the REST API is worth doing, then say so...and start opening new Issues to track the implementation. Otherwise, you can close this as an Issue.

**Comment by jbholden** [comment]
Making a REST API to load the past years is probably not worth doing.  It was helpful last year when having to load the games and results for each week from a spreadsheet, but if the spreadsheets aren't going to be used anymore then it's not worth it.
